FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,136 @@
+//
+//  Created by Nazar Alwi on 27/05/21
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import CoreData
+
+public final class CoreDataFeedStore: FeedStore {
+    private let container: NSPersistentContainer
+    private let context: NSManagedObjectContext
+    
+    public init(storeURL: URL, bundle: Bundle = .main) throws {
+        container = try NSPersistentContainer.load(modelName: "FeedStore", url: storeURL, in: bundle)
+        context = container.newBackgroundContext()
+    }
+    
+    public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+        perform { context in
+            do {
+                try ManagedCache.find(in: context).map(context.delete).map(context.save)
+                completion(nil)
+            } catch {
+                completion(error)
+            }
+        }
+    }
+    
+    public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+        perform { context in
+            do {
+                let managedCache = try ManagedCache.newUniqueInstance(in: context)
+                managedCache.timestamp = timestamp
+                managedCache.feed = ManagedFeedImage.images(from: feed, in: context)
+                
+                try context.save()
+                completion(nil)
+            } catch {
+                completion(error)
+            }
+        }
+    }
+    
+    public func retrieve(completion: @escaping RetrievalCompletion) {
+        perform { context in
+            do {
+                if let cache = try ManagedCache.find(in: context) {
+                    completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
+                } else {
+                    completion(.empty)
+                }
+            } catch {
+                completion(.failure(error))
+            }
+        }
+    }
+    
+    private func perform(_ action: @escaping (NSManagedObjectContext) -> Void) {
+        let context = self.context
+        context.perform { action(context) }
```

Comment:
During the course this way of having a local reference to the context to avoid capturing self inside of the `context.perform` block is taught.
The same can be achieved using capture list.
It saves one line and one variable.

No need to make this amendment, this is just for your interest.

```suggestion
        context.perform { [context] in action(context) }
```